### PR TITLE
API improvements

### PR DESCRIPTION
**Pull request by rowanseymour**
* QuerySet counts used for pagination totals are cached by list views
* Upgraded to latest Django REST framework 2.x to get fix for duplicated queries bug
* Reworked results endpoint to not require a "fake" serializer
* Removed unused form rendering code from API
* Cleaned up serializers (got rid of unused `read_only_fields`, tidied up user/org args)
* Made API endpoint naming consistent, e.g. _Something_Endpoint
* GET requests to API endpoints are no longer atomic (i.e. transaction no longer required)
* Removed duplicate listing for broadcasts GET/POST in the API explorer
* Optimized the contacts endpoint so that it doesn't make repeated db calls for each returned contact
* Improved performance tests which can assert number of queries and/or transactions
* Reworked all write serializers to inherit from `WriteSerializer` so they provide helpful error message when post body is not a JSON object
* Stopped endpoints from pointlessly fetching data when being browsed as HTML
* Added unit tests for using API as XML and browsing API endpoints as HTML

### CODE REVIEW

File: pip-freeze.txt [hunk 1]
```diff
@@ -22,7 +22,7 @@ BeautifulSoup==3.2.1
 xlutils==1.7.1
 phonenumbers==7.0.2
 django-countries==3.0.1
-djangorestframework==2.2.1
+djangorestframework==2.4.5
```
Comment (rowanseymour) [inline pip-freeze.txt]: There's a bug in 2.2.1 which duplicates queries for paginated requests. This is the latest 2.x release.

File: temba/api/channels.py [hunk 2]
```diff
@@ -0,0 +1,1265 @@
+from __future__ import absolute_import, unicode_literals
```
Comment (rowanseymour) [inline temba/api/channels.py]: Moved all the channel handlers in api/views.py into their own module as the former was getting unwieldy

File: temba/api/renderers.py [hunk 3]
```diff
@@ -1,80 +0,0 @@
-from __future__ import unicode_literals
```
Comment (rowanseymour) [inline temba/api/renderers.py]: We're not using any of this as far as I can tell. We have browsable html versions of our API endpoints but they don't include forms - just documentation.

File: temba/api/serializers.py [hunk 4]
```diff
@@ -141,8 +140,8 @@ class MsgBulkActionSerializer(serializers.Serializer):
     label_uuid = serializers.CharField(required=False)
 
     def __init__(self, *args, **kwargs):
-        if 'user' in kwargs:
-            self.user = kwargs.pop('user')
+        self.user = kwargs.pop('user')
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Our serializers now always take a `user` arg, and no `org` arg. Trying to be consistent about how we use user and org in all of our serializers

File: temba/api/serializers.py [hunk 5]
```diff
@@ -308,20 +303,22 @@ class ContactReadSerializer(serializers.ModelSerializer):
...
-        return [_.uuid for _ in obj.user_groups.all()]
+        groups = obj.prefetched_user_groups if hasattr(obj, 'prefetched_user_groups') else obj.user_groups.all()
+        return [_.uuid for _ in groups]
 
     def get_urns(self, obj):
         if obj.org.is_anon:
             return dict()
 
-        return [urn.urn for urn in obj.urns.all()]
+        return [urn.urn for urn in obj.get_urns()]
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Will use cached URNs if available

File: temba/api/serializers.py [hunk 6]
```diff
@@ -691,20 +679,19 @@ class CampaignEventWriteSerializer(serializers.Serializer):
...
-            self.user = kwargs['user']
-            del kwargs['user']
+        self.user = kwargs.pop('user')
+        self.org = self.user.get_org()
 
         super(CampaignEventWriteSerializer, self).__init__(*args, **kwargs)
 
     def validate_campaign(self, attrs, source):
-        if not source in attrs:
+        if source not in attrs:
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Previous syntax is a PEP warning

File: temba/api/serializers.py [hunk 7]
```diff
@@ -1008,31 +987,6 @@ def initialize(self, parent, field_name):
         super(FlowField, self).initialize(parent, field_name)
 
 
-class ResultSerializer(serializers.Serializer):
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Latest Django REST framework doesn't seem to like this as it's not really a serializer in that there's no object or model. I've moved this functionality into the view.

File: temba/api/views.py [hunk 8]
```diff
@@ -260,34 +256,31 @@ def get_context_data(self, **kwargs):
...
-        endpoints.append(BroadcastsEndpoint.get_write_explorer())
+        endpoints.append(BroadcastEndpoint.get_read_explorer())
+        endpoints.append(BroadcastEndpoint.get_write_explorer())
 
-        endpoints.append(LabelsEndpoint.get_read_explorer())
-        endpoints.append(LabelsEndpoint.get_write_explorer())
+        endpoints.append(LabelEndpoint.get_read_explorer())
+        endpoints.append(LabelEndpoint.get_write_explorer())
 
-        endpoints.append(BroadcastsEndpoint.get_read_explorer())
```
Comment (rowanseymour) [inline temba/api/views.py]: Broadcasts GET/POST was being listed twice on the API explorer page

File: temba/api/serializers.py [hunk 9]
```diff
@@ -677,7 +666,6 @@ def get_relative_to(self, obj):
     class Meta:
         model = CampaignEvent
         fields = ('event', 'campaign', 'relative_to', 'offset', 'unit', 'delivery_hour', 'message', 'flow', 'created_on')
-        read_only_fields = ('offset', 'unit', 'message', 'delivery_hour', 'created_on')
```
Comment (rowanseymour) [inline temba/api/serializers.py]: All our API views use separate write serializers so `read_only_fields` doesn't make sense here

File: temba/api/serializers.py [hunk 10]
```diff
@@ -11,15 +11,19 @@
...
-from temba.flows.models import Flow, FlowRun, RuleSet
+from temba.flows.models import Flow, FlowRun
 from temba.locations.models import AdminBoundary
-from temba.msgs.models import Msg, Call, Broadcast, Label, VISIBLE, ARCHIVED, DELETED
-from temba.values.models import Value, VALUE_TYPE_CHOICES
+from temba.msgs.models import Msg, Call, Broadcast, Label, ARCHIVED
+from temba.values.models import VALUE_TYPE_CHOICES
 
 
+# ------------------------------------------------------------------------------------------
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Just putting all fields in their own section so it's easier to find them

File: temba/api/serializers.py [hunk 11]
```diff
@@ -11,15 +11,19 @@
...
+# Field types
+# ------------------------------------------------------------------------------------------
+
 class DictionaryField(serializers.WritableField):
 
-    def to_native(self, obj):  # pragma: no cover
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Added tests for these so `#pragma: no cover` no longer appropriate

File: temba/api/tests.py [hunk 12]
```diff
@@ -182,21 +261,25 @@ def test_api_flows(self):
...
+        # login as plain user
         self.login(self.user)
         self.assert403(url)
 
-        # log in a manager
+        # login as administrator
         self.login(self.admin)
 
         # test that this user has a token
-        self.assertTrue(self.admin.api_token)
```
Comment (rowanseymour) [inline temba/api/tests.py]: `api_token` is a method so this wasn't testing anything

File: templates/rest_framework/api.html [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {% extends "frame.html" %}
 
-{% load rest_framework %}
+{% load rest_framework static %}
```
Comment (rowanseymour) [inline templates/rest_framework/api.html]: Needed this after REST framework upgrade

File: templates/rest_framework/form.html [hunk 14]
```diff
@@ -1,22 +0,0 @@
-{% load rest_framework smartmin %}
```
Comment (rowanseymour) [inline templates/rest_framework/form.html]: File not being used

File: temba/api/channels.py (hunk 2 above)
Comment (nicpottier) [inline temba/api/channels.py]: Yay for that.

File: temba/api/tests.py (hunk 12 above)
Comment (nicpottier) [inline temba/api/tests.py]: Shouldn't we just call it then?

File: temba/api/views.py [hunk 15]
```diff
@@ -408,13 +402,116 @@ def api(request, format=None):
...
+
+            cached_count = cache.get(count_key)
+            if cached_count is not None:
+                queryset.fixed_count = int(cached_count)
+
+            page = super(ListAPIMixin, self).paginate_queryset(queryset)
+
+            if cached_count is None:
+                calculated_count = page.paginator.count
+                cache.set(count_key, calculated_count, REQUEST_COUNT_CACHE_TTL)
```
Comment (nicpottier) [inline temba/api/views.py]: Don't we want to reset the expiration either way?  IE, paging across a few million messages the count "should" be consistent? Cache should expire 5 mins after the LAST page was refetched, not the first page.

**Comment by nicpottier** [comment]
Nice, that should seriously help.

File: temba/api/views.py (hunk 15 above)
Comment (rowanseymour) [inline temba/api/views.py]: Not so sure. Currently there's no consistency.. obviously new items can be inserted as you go. Also I think a fairly high percentage of API calls are paged but without all pages being fetched. For example our infinite scrolling in CasePro uses paging and only requests new pages of messages if a user scrolls down. When we cache that count we have no idea if the last page will ever be requested.

Comment (nicpottier) [inline temba/api/views.py]: By last page I don't mean the last page in the result set, just the last page requested. IE, if you stop paginating after page 100, then the key will expire five minutes after that. But if you spend three hours paginating over 100,000 pages, you'll get consistent counts throughout, which I think makes a lot more sense to explain to someone. (counts are determined on the first query and cached thereafter)

Comment (rowanseymour) [inline temba/api/views.py]: Hmm. We'd have to sure that repeated calls with the same params can't lead to the cache _never_ expiring (like when that org was polling for all their runs with 2 simultaneous threads). So I think we'd force a re-calculation when you request page 1.

Comment (nicpottier) [inline temba/api/views.py]: Oh ya for sure, should always recalculate on the first page.

**Pull request merged**